### ISSUE TITLE
Provide the ability to trust self-signed certificates by providing a custom CA

### ISSUE THREAD

**Issue by lostiniceland** [prisma/tiberius#197]
As mentioned in #176, tiberius currently does not provide the option to register a custom CA certificate.
When the server is running with a self-signed cert, currently the only option is to trust any certificate, which basically disables the trust completely.

The tiberius config should be enhanced with an option to register a custom CA which is then passed to the underlying ssl implementation.

**Comment by pimeys** [comment]
The code is going to be very easy. The hard part here is to get a server with a custom cert, and that cert file. Then just see whatever for example the JDBC implementation takes as the connection string parameter for the path and implement the public API for Tiberius.

If you have already the testing environment to try it out, I can provide guidance to get the code done.

Where you want to start looking is in the [connection.rs](https://github.com/prisma/tiberius/blob/main/src/client/connection.rs#L384): we use the `TlsConnector` from async-native-tls (read the [docs](https://github.com/prisma/tiberius/blob/main/src/client/connection.rs#L384)), which has methods to take a custom `Identity` and `Certificate` in. Read the docs and it should be quite straightforward.

The question then is what would be the Tiberius interface for getting the file paths. I'd say whatever the JDBC or dotnet drivers do with their connection strings works for Tiberius too.

The optimal thing would be to have a CI setup for custom certs, and we already run tests on Windows which might be easier to setup with custom certs.

**Comment by pimeys** [comment]
Oh and digging deeper into client/connection/tls here is not super important, there's lots of bullshit there due to how the Microsoft TLS handshake goes. None of that should be needed to finish this feature.

**Comment by lostiniceland** [comment]
Thanks for the update. I started this morning setting up the test-environment and integrating it in the linux test-pipeline.
Lets see how far I get. My time is very limited so it might take a while.

**Comment by lostiniceland** [comment]
@pimeys I've got a working solution. Since I did not want to change existing APIs, there is now a 'trust_ca(path)' next to `trust_cert()' which is internally mapped to a [new enum type](https://github.com/lostiniceland/tiberius/commit/0e9a98865f33b9f25e772b9035bacb6fdd0e51df#diff-b19759681d268fddf3754d17d98c22f273430fae1ecb3cb24b8d3be57a1baed7R38-R42).

Let me know what you think.

What remains open, is the proper integration into the test-pipeline. Furthermore I am not sure if this feature should be available as a connection-string-
Last but not least, I am getting a lot of unit-test error, but those happen on main as well. Maybe you can guide me to the proper way running all tests locally.

**Comment by pimeys** [comment]
So, do not enable gssapi stuff if you don't have the library installed. In general the development workflow here is with nix, so you can read the `flake.nix` to see what's needed. Or even better, install nix and just run the `nix develop` to get everything you need to work with Tiberius.

But in your case, I'd just try to run `cargo test` and disable the kerberos functionality.

Also, I was quickly checking your code and it is definitely worth a PR soon. I don't think we need CI for this feature, but one I'd like still to see in the PR is supporting the certs in the JDBC and ADO.net connection strings. The best would be to directly mimic whatever the official Microsoft clients do.

P.S. Yeah I know I should document everything even better here. I'm actually also thinking of writing an article about how Tiberius works, what is needed with the TDS protocol and how certain things were solved here due to the limitations of the protocol. Just need some time...

**Comment by lostiniceland** [comment]
According to [this documentation for ADO.NET](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/connection-string-syntax) there is no option to pass a CA. Also the [JDBC docs](https://docs.microsoft.com/en-us/sql/connect/jdbc/client-certification-authentication-for-loopback-scenarios?view=sql-server-ver15) only mention connecting via client-certificate.

Granted, usually the CA should be setup in the systems trust-chain, but during development it is sometimes useful to just setup some custom-CA.

So I would have to come up with some custom connection-string key to support this. Or, just keep it on the API layer.

**Comment by pimeys** [comment]
I would be OK with the JDBC key; having it the same in the ADO.net. In Prisma, the connection string is the only interface the user has to Tiberius. This feature might be useful for some of them too...

**Comment by lostiniceland** [comment]
I've opened a PR. In order to integrate the tests, I enabled the custom-ca for the existing docker-compose setups (instead of adding another one), because all integration-tests already ran with `TrustServerCertificate=true`. The new integration-test is only enabled on Unix.
Lets see what the pipeline says...

**Comment by pimeys** [comment]
Cool. Make the CI happy first, I'll do a review on code and we'll cut a new release!

**Comment by lostiniceland** [comment]
I am currently trying to replicate the pipeline test-failures locally, but I am still facing this gssapi error.
I've setup a nix environment which I start with `nix develop --extra-experimental-features nix-command --extra-experimental-features flakes` (the extra flags were obviously necessary...at least then it started).

Running `cargo test --features=all` in the nix environment gives a lot of the following errors 

`Error: Gssapi("gssapi major error Unspecified GSS failure.`

Since I got the same error-count on main, I thought the PR should be good to go.
Could you assist me a bit more in whats causing this nix problem?

**Comment by pimeys** [comment]
I don't think you need to worry about the kerberos, just test without:

```bash
cargo test --features chrono,tds73,sql-browser-tokio,sql-browser-async-std,sql-browser-smol,rust_decimal,bigdecim
al
```

I see the failing tests of interest in CI are trying to connect to the server without TLS at all. Maybe check them out:

```

---- connect_as_plain_text_async_std stdout ----
Error: Tls("unexpected EOF")
thread 'connect_as_plain_text_async_std' panicked at 'assertion failed: `(left == right)`
error: test failed, to rerun pass '--test query'
  left: `1`,
 right: `0`: the test returned a termination value with a non-zero status code (1) which indicates a failure', /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/test/src/lib.rs:194:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- connect_as_plain_text_tokio stdout ----
Error: Tls("unexpected EOF")
thread 'connect_as_plain_text_tokio' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`: the test returned a termination value with a non-zero status code (1) which indicates a failure', /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/test/src/lib.rs:194:5
```

To get the formatting failures to go away, run `cargo fmt`. This is handy to do always when saving a file in your editor.

**Comment by lostiniceland** [comment]
Just fixed another crux with the local test-setup. Most of the test failed locally because no user/password was provided in the connection-string. Though this was provided via env in Github actions. So basically the tests cannot be run unless you overwrite the connection-string via env. At least now I have the same result as the CI :-)
Should I fix this on the way or is there a reason for this?

**Comment by pimeys** [comment]
Released in 0.7.2.
